File: src/main/java/datastructures/queue/java/QueueWithTwoStacks.java
package datastructures.queue.java;

import datastructures.stack.kotlin.CustomStack;

public class QueueWithTwoStacks<T> {
    private int count;
    private CustomStack<T> primaryStack;
    private CustomStack<T> secondaryStack;
    private int size;

    public QueueWithTwoStacks(int size) {
        this.primaryStack = new CustomStack<>(5);
        this.secondaryStack = new CustomStack<>(5);
        this.size = size;
    }

    public void enqueue(T item){
        if(count == size) throw  new IllegalStateException();
        primaryStack.push(item);
        count++;
    }

    public void dequeue(){
        if(isEmpty()) throw new IllegalStateException();
        movePrimaryStackDataToSecondary();
        secondaryStack.pop();
        count --;
    }

    public Object peek(){
        if(isEmpty()) throw new IllegalStateException();
        movePrimaryStackDataToSecondary();
        return secondaryStack.peek();
    }

    private void movePrimaryStackDataToSecondary(){
        if(secondaryStack.isEmpty()){
            while (!primaryStack.isEmpty()){
                secondaryStack.push((T) primaryStack.pop());
            }
        }
    }

    public boolean isEmpty() { return secondaryStack.isEmpty() && primaryStack.isEmpty(); }
}

File: src/main/java/datastructures/queue/kotlin/QueueWithTwoStacks.kt
package datastructures.queue.kotlin

import datastructures.stack.kotlin.CustomStack

class QueueWithTwoStacks<T>(val size : Int) {
    private var primaryStack: CustomStack<T> = CustomStack(size)
    private var secondaryStack: CustomStack<T> = CustomStack(size)


    fun enqueue(item : T){
        primaryStack.push(item)
    }

    fun dequeue() : Any?{
        moveStackData();
        return secondaryStack.pop();
    }

    fun peek() : Any?{
        moveStackData();
        return secondaryStack.peek();
    }

    private fun moveStackData(){
        if(secondaryStack.isEmpty()){
            while (!primaryStack.isEmpty()){
                secondaryStack.push(primaryStack.pop() as T);
            }
        }
    }

    fun isEmpty() : Boolean { return  primaryStack.isEmpty() && secondaryStack.isEmpty() }
}